src/main/java/org/opensearch/sdk/Extension.java

```diff
@@ -84,4 +85,14 @@ default Collection<Object> createComponents() {
     default List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
         return Collections.emptyList();
     }
+
+    /**
+     * Provides the list of implimented interfaces by an extension, empty if
+     * none.
+     *
+     * @return list of Interfaces implimented by extension
+     */
+    default List<ExtensionsManager.ExtensionInterfaceType> getImplimentedInterfaces() {
```

Comment:
We don't need a separate extension point for this.  The extension will have done `implements Extension, ActionExtension, ... ` so Java has this information available to it.

A brute force way to do it is just a long set of `if (extension instanceof ActionExtension) { .... add ACTION to enumset ... }`.   

Another option (I'll refrain from saying it's more elegant) is to iterate over `extension.getClass().getInterfaces()` and do a little bit of parsing to translate the class type to the enum (truncate the before the word `Extension` and [convert camelcase to underscorecase](http://www.java2s.com/example/java-utility-method/string-camel-case-to-underscore/camelcasetounderscores-string-camel-02556.html)).

Reply:
Yes yes, the second option is elegant. I will implement that way.

Reply:
> Yes yes, the second option is better. I will implement that way.

So as usual in the process of going through code we find a better way.  I am also thinking all this parsing of class name to enum isn't really needed either.  Ultimately all we want to do is an `implementedExtensions.contains(interfaceName)` and we can do that with a string as easily as an enum.  

Which means you could just take that `getInterfaces()` result and pass it.  (Probably don't need the package name prefix on it.).  So we'd just do `implementedExtensions.contains("ActionExtension")` when we wanted to check that, for example.

Reply:
And thinking more on this, I see a flaw in that if someone implements their own interface that's not part of the SDK but has the same name, it'd trip us up unless we used the fully qualified class name.  Which ends up being a lot more big long strings being passed around everywhere, and I'm thinking the brute force `instanceof` approach may be the best.

Reply:
Hi @dbwiddis ,

Yes these will create problem. Hadn't thought of it before. But what if we just check whether Interfaces belonging to org.opensearch.sdk and only send those ? ( I am imagining developer who builds the extension would have placed his interface in different package ). 

Yes now that I think of it, List<String> is better than List<Enum> as even after we store, while comparing, it will be trouble. So I have changed it to String  so we can just use .contains() on implementedInterfaces of DiscoverExtension.

I will commit these and I have added one testcase as well. Tell me what else can be improved.

Thanks and Regards
Varun

Reply:
As of now I have modified the HelloWorldExtension class, but for testcases may be I should define some private extension classes which impliments Engine and Search Extension ? ( we can add more test scenarios )

Thanks and Regards
Varun

Reply:
Sorry, trying to track both PRs and not sure where we are or if you're waiting on my review.

I think we're going with a list (or set) of String interface names, filtering with `org.opensearch.sdk` prefix?

Reply:
Hi @dbwiddis ,

Yes currently I have implemented like this. I can change to method of just checking with **instaceof** approach if you like. And I have updated the other PR as well. I believe once both of the PR gets ready, We can update on both the side simultaneously as we are modifying ExtensionInitResponse method structure. 

I will change HelloWorldExtension, I had just changed it just to test once. Apart from it Any other changes you want me to do ? ( In both here as well as other PR )

Also, I did not understand what you meant in the first comment review **Comments inline!**.

Thanks and regards
Varun.

Reply:
OK, looks good, I like this approach!  Reviewed the companion PR as well.   Some suggestions:
 - rather than package name string equality use `equals` or possibly a `startsWith` as we may move the extension interfaces into subpackages and that string wouldn't auto-update.  Might even be good to get the package location of one of the known-good interfaces (`Extnsion`) automatically instead of hardcoding a string, so as long as all the extension interfaces are in the same package, anywhere, it works.
 - for the `Writeable` implementations, there are `readStringArray` and `writeStringArray` methods that do the same logic you are with the vInt, and you can go to/from list to array easily. I'd suggest that.
 - The openSearch side will need the test case improved for the class you're editing

By "comments inline" I meant this comment thread we're on, rather than giving a summary comment. It's usually used in email when you add comments in the middle of the text you're replying to, and adding comments in the middle of code is kind of the same thing.  You can ignore it.

I'm on vacation until next Sunday. Will probably poke around on these PRs in the evening but responses will be slower than usual.  Thanks for all the work you're doing.

Reply:
Hi @dbwiddis,

- Yes good point I will replace the current one with startswith. That will be better.
- I changed both writeTo and constructor to use readStringArray and writeStringArray.
- I looked at the testcase of ExtensionsManager (TestExtensionsManager), the method we are interested in is **testInitialize**, where current implimentation is they just create ExtensionsManager instance with the yml file which they have defined and they try to initalize, and they just assert the exception logs which they get as initialize will try to send request to sdk node ( they have not mocked it yet ). there is one issue raised on behalf of it. should I try to resolve this https://github.com/opensearch-project/OpenSearch/issues/4045? only then I can actually test the interfaces coming from the response.

Reply only if you get time and enjoy the vacation !!. I might need some time to write this testcase. 

Thanks and regards
Varun.

Reply:
Hmm, looks like that `InitializeExtensionResponse` (and the corresponding `InitializeExtensionRequest` got missed in unit testing.   Would be nice to either add a new test class in `org.opensearch.discovery` class to test these, or add a test for them in the `ExtensionManagerTests` class where there are plenty of examples exercizing the writable streaminput / streamoutput stuff.

Reply:
Yes, I will write unit tests for both and that for extensionsmanager as well.

Thanks
Varun

Reply:
Hi @dbwiddis ,

I have done all the changes. I have taken actionExtension Interface as a reference to compare and I have used .startswith() function and I have added a testcase for InitExtensionResponse ( where I have changed ). InitExtensionRequest involves transportService which I need to understand better to write it. Testcase of initialize of ExtensionsManager also needs more time ( Since I need to mock sdk node ). I thought I will come to these again later ( Especially the one with mocking sdk node ).

Tell me if its okay else I can spend more time for writing testcase for InitExtensionRequest if its necessary. Also tell me if any other changes needs to be done.

Thanks and Regards
Varun.